### PR TITLE
feat/filter by expiration and sorting

### PR DESCRIPTION
**Pull request by maycon-mello**
- Filter by expiration date
- Sorting

### CODE REVIEW

File: src/features/credentials/filters/components/FilterExpirationDate.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+          });
+        }}
+        placeholder={'Select a value'}
+        _selectedItem={{
+          bg: 'teal.600',
+          endIcon: <CheckIcon size={5} />,
+        }}
+        mt="1">
+        <Select.Item key={0} label={'Select a value'} value={null} />
+        <Select.Item key={1} label={'Hide expired'} value={'notExpired'} />
```
Comment (mike-parkhill) [inline src/features/credentials/filters/components/FilterExpirationDate.tsx:27]: I was expecting this to be a date range filter (i.e. show creds that expired between date 1 and date 2), but this is probably a good start

we can add the date range later if people actually ask for it

Comment (maycon-mello) [reply src/features/credentials/filters/components/FilterExpirationDate.tsx:27]: Sounds good. I've added support for date range in the filtering function, would need to update the UI only.

Comment (maycon-mello) [reply src/features/credentials/filters/components/FilterExpirationDate.tsx:27]: just added a unit tests for that

File: src/features/credentials/filters/components/SortingDropdown.tsx [hunk 2]
```diff
@@ -0,0 +1,89 @@
+import {SortingDirection, SortingOptionID} from '../filter-types';
+import {Box, CheckIcon, FormControl, Select} from 'native-base';
+import React from 'react';
+import {translate} from '../../../../locales';
+
+const SORTING_OPTIONS: SortingOptionID[] = [
+  'issuanceDate',
+  'issuer',
+  'expirationDate',
+  'holder',
```
Comment (mike-parkhill) [inline src/features/credentials/filters/components/SortingDropdown.tsx:10]: should we display the holder/subject ID on the preview card so this sort makes more sense?

maybe issuer name or DID, too?

Comment (maycon-mello) [reply src/features/credentials/filters/components/SortingDropdown.tsx:10]: I agree, can we add this as part of https://dock-team.atlassian.net/browse/CM-964?

File: src/features/credentials/filters/components/SortingDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  onChange,
+  filter,
+}: {
+  onChange: (filter: any) => void;
+  filter: any;
+}) {
+  return (
+    <Box mb={2}>
+      <FormControl w="100%" isInvalid px={5}>
+        <FormControl.Label>Sort by</FormControl.Label>
```
Comment (mike-parkhill) [inline src/features/credentials/filters/components/SortingDropdown.tsx:24]: are you still waiting on UX designs before moving these labels to translation files?

Comment (maycon-mello) [reply src/features/credentials/filters/components/SortingDropdown.tsx:24]: Yes

File: src/features/credentials/filters/credential-filters.tsx [hunk 4]
```diff
@@ -108,3 +113,133 @@ export function filterByText({credentials, filter}: FilterFnInput) {
...
+
+      if (direction === 'asc') {
+        return aDate - bDate;
+      }
+
+      return bDate - aDate;
+    });
+  }
+
+  if (id === 'expirationDate') {
```
Comment (mike-parkhill) [inline src/features/credentials/filters/credential-filters.tsx:203]: similar to the control map for the filters, would be nice to make the sort rules pluggable too.

File: src/features/credentials/filters/use-credential-filters.test.ts [hunk 5]
```diff
@@ -171,4 +183,269 @@ describe('useCredentialFilters', () => {
...
+        }) as any,
+        createMockCredential({
+          expirationDate: moment().add(10, 'days').toDate().toString(),
+        }) as any,
+        createMockCredential({
+          expirationDate: undefined,
+        }) as any,
+      ];
+    });
+    it('expect to not return expired credentials', () => {
```
Comment (mike-parkhill) [inline src/features/credentials/filters/use-credential-filters.test.ts:203]: dateRange test?

Do you have a test where no expiration filter is specified but the sample set has both expired and non-expired creds so we can verify that all creds are returned regardless of expiration?

Comment (maycon-mello) [reply src/features/credentials/filters/use-credential-filters.test.ts:214]: Done, test was added

File: src/features/credentials/filters/credential-filters.tsx [hunk 6]
```diff
@@ -108,3 +113,133 @@ export function filterByText({credentials, filter}: FilterFnInput) {
...
+    return credentials;
+  }
+
+  const filterInput: ExpirationDateRange = filter.value as ExpirationDateRange;
+
+  if (!filterInput) {
+    return credentials;
+  }
+
+  if (filterInput.mode === 'notExpired') {
```
[the remark this answers is not in the document]
Comment (maycon-mello) [reply src/features/credentials/filters/credential-filters.tsx:128]: This is meant to be an alias. The string type has better reusability allowing us to added new modes quickly, such as: 'notExpired' | 'expired' | 'expiring' | 'dateRange'. 

Maybe I can rename `notExpired` to `valid`

File: src/features/credentials/filters/credential-filters.tsx (hunk 4 above)
Comment (maycon-mello) [reply src/features/credentials/filters/credential-filters.tsx]: fixed

File: src/features/credentials/filters/components/SortingDropdown.tsx (hunk 2 above)
Comment (maycon-mello) [reply src/features/credentials/filters/components/SortingDropdown.tsx]: done

File: src/features/credentials/components/CredentialCard/CredentialCard.tsx [hunk 7]
```diff
@@ -128,6 +132,38 @@ export function CredentialCard({
...
+                fontSize={'13px'}
+                fontWeight={600}
+                fontFamily={Theme.fontFamily.montserrat}>
+                Issuer:
+              </Typography>
+              <Typography
+                fontSize={'13px'}
+                fontWeight={300}
+                fontFamily={Theme.fontFamily.montserrat}>
+                {credential?.issuer?.id}
```
Comment (mike-parkhill) [inline src/features/credentials/components/CredentialCard/CredentialCard.tsx:147]: name might be more meaningful here

Comment (mike-parkhill) [reply src/features/credentials/components/CredentialCard/CredentialCard.tsx:147]: if we have it...

so maybe issuer?.name || issuer?.id or something

File: src/features/credentials/components/CredentialCard/CredentialCard.tsx [hunk 8]
```diff
@@ -128,6 +132,38 @@ export function CredentialCard({
...
+                fontSize={'13px'}
+                fontWeight={600}
+                fontFamily={Theme.fontFamily.montserrat}>
+                Holder:
+              </Typography>
+              <Typography
+                fontSize={'13px'}
+                fontWeight={300}
+                fontFamily={Theme.fontFamily.montserrat}>
+                {credential?.credentialSubject?.id}
```
Comment (mike-parkhill) [inline src/features/credentials/components/CredentialCard/CredentialCard.tsx:163]: again, probably better UX to show the name if it exists

**Comment by mike-parkhill** [comment]
These implementations will have to change a fair bit when we settle on a different data store to optimize for larger credential sets, but for now this approach will work.

File: src/features/credentials/components/CredentialCard/CredentialCard.tsx (hunk 8 above)
Comment (maycon-mello) [reply src/features/credentials/components/CredentialCard/CredentialCard.tsx]: fixed

File: src/features/credentials/components/CredentialCard/CredentialCard.tsx (hunk 7 above)
Comment (maycon-mello) [reply src/features/credentials/components/CredentialCard/CredentialCard.tsx]: fixed

**Comment by maycon-mello** [comment]
> These implementations will have to change a fair bit when we settle on a different data store to optimize for larger credential sets, but for now this approach will work.

Yep, I was planning on using typeorm (sqlite). We can still using the same structure for filters, but instead of passing a credential list as parameter for each filter, we could pass a queryBuilder.  something like: 

```js
queryBuilder.where('credential.type = :type', { type: 'BasicCredential' });
queryBuilder.andWhere('credential.issuanceDate > :date', { date:'some-date' });

```

I think unit tests would keep the same

**Review by mike-parkhill — DISMISSED** [review]

**Review by mike-parkhill — APPROVED** [review]

**Pull request merged**